DECENTRALIZED_DATA_DELIVERY_MARKETS.md Open Problem Statement (#14)
protocol/ResNetLab: OPEN_PROBLEMS/DECENTRALIZED_DATA_DELIVERY_MARKETS.md

    *   How to verify that the file being transferred is the one requested, before a large amount of resources has been invested (i.e., before the entirety of the file has been transmitted)
    *   How to verify that the client has received the file
  *   How to verify and reward that SLA were guaranteed?
  *   How to avoid collusion when adding a third-party (e.g. Referee)?
  *   How to avoid grieving attacks (make one party pay for fees)?
  *   How to avoid a malicious actor causing un-rewarded work, hence wasting the others’ resources?
*   Experience
  *   How to make the transfers start instantaneously?
  *   How to support others (e.g. Content Publishers) paying for the usage?
  *    How to make it private (e.g. that others don’t know who is requesting what)?
*   Performance
  *   How to overcome the send-and-halt in order to max bandwidth throughput 
      *   How to support multipath (i.e. fetching from multiple sources)

### State-of-the-art

When it comes to the state of the art, we found that the existing solutions fall within one of the following categories:

> #### Pay-per-packet
> 
> INSERT DESCRIPTION
> 
> [(David) Original Filecoin payment channels](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.pc4bfmnw2ke0)	[4](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.pc4bfmnw2ke0)
> [(David) Theta Whitepaper](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.82adp2bb1rys)	[4](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.82adp2bb1rys)
> 
> **Known shortcomings of these approaches:**
> *   TOFILL
> 

Author: ```suggestion
#### Pay-per-packet

Pay-per-packet solutions offer granular control over what gets paid, enabling the option to verify if the SLA is being fulfilled. However, their setup cost and latency overhead puts them behind on being able to max out the available underlay resources.

These type of solutions have been found/highlighted:

- [Filecoin Retrieval Market](https://spec.filecoin.io/#section-systems.filecoin_markets.retrieval_market)
- [Theta Whitepaper](https://s3.us-east-2.amazonaws.com/assets.thetatoken.org/Theta-white-paper-latest.pdf?v=1612955013.791)

**Known shortcomings of these approaches:**
*   send-and-halt - Because the next packet (unit of service) will only be issued after the payment of the previous, these solutions often miss to max out the bandwidth available in the connection available, increasing the latency of delivery.

```

# Decentralized Data Delivery Markets (3DMs) - Open Problem Statement

> DO NOT MERGE WITHOUT CREATING DOCTOC https://www.npmjs.com/package/doctoc

Author: - [ ] Do before merge

# Overall

## Short description

> With the emergence of Decentralized Storage Networks and the rapid decrease in the price of storage services and hardware, there is a rapidly growing need to leverage the additional storage capacity contributed to Decentralized Storage Networks by several new players, such as end-users and use it to deliver a reliable and high quality storage but also delivery services. Similarly to Content Delivery Networks (CDNs) for the traditional Cloud Storage market, we now have the opportunity to build **Decentralised CDNs **for Decentralised Storage Networks. The Decentralized Data Delivery Markets (3DMs) Open Problem covers all the essential areas of work that need to be studied in order to create **a fully permissionless free market for data delivery that supports fair data exchange** on the service provided.

Reviewer: ```suggestion
With the emergence of Decentralized Storage Networks and the rapid decrease in the price of storage services and hardware, there is a rapidly growing need to leverage the additional storage capacity contributed to Decentralized Storage Networks by new players, including end-users, and use it to deliver reliable and high-quality storage and delivery services. Similarly to Content Delivery Networks (CDNs) for the traditional Cloud Storage market, we now have the opportunity to build **Decentralised CDNs** for Decentralised Storage Networks. The Decentralized Data Delivery Markets (3DMs) Open Problem covers all the essential areas of work that need to be studied in order to create **a fully permissionless free market for data delivery that supports fair data exchange** on the service provided.
```

## Long description 

> Serving content globally at scale is a hard technical problem as evidenced by the multiple decades of innovation and improvements in the Content Delivery Networks field. This challenge becomes even more interesting once we move away from centralized cloud infrastructures, which is typically managed and monitored by a single party, to a decentralized network that is permissionless, likely anonymous, constantly changing (i.e. high node churn) and without access to the convenience of a third party mediator system that facilitates the fair exchange of goods (e.g. credit providers). The benefits of moving to a decentralised setting are significant, however: cheaper storage and delivery services, resilience against business failures (i.e., the network and all its components are not dependent on business decisions made by a single entity), moving away from the personal data-based business models, as well as a significantly lower barrier to entry for new players who want to contribute to the system.

Reviewer: ```suggestion
Serving content globally at scale is a hard technical problem, as evidenced by the multiple decades of innovation and improvements in the Content Delivery Networks field. This challenge becomes even more interesting once we move away from centralized cloud infrastructure, which is typically managed and monitored by a single party, to a decentralized network that is permissionless, possibly anonymous, constantly changing (i.e. with high node churn) and lacking access to the convenience of a third party mediator system that facilitates the fair exchange of goods (e.g. credit providers). The benefits of moving to a decentralised setting are significant, however: cheaper storage and delivery, resilience against business failures (i.e., the network and all its components are not dependent on business decisions made by a single entity), independence from the personal data-driven business models, as well as a significantly lower barrier to entry for new players who want to contribute.
```

> At the same time, common problems addressed in the past, in traditional CDNs, reappear, albeit in a different dimension as we move to a decentralized network design space. Some of these problems are: the global orchestration of the system, the efficient allocation and use of the resources, or a clear visibility of the content being retrieved from the network.

Reviewer: ```suggestion
At the same time, problems already addressed in the past for traditional CDNs reappear, albeit in a different dimension, as we move to a decentralized network design space. These problems include the global orchestration of the system, the efficient allocation and use of the resources, and clear visibility of the content being retrieved from the network.
```

Reviewer: The phrasing here is a little unclear: do we consider the content being visible a problem (i.e. is this a privacy issue?) or do we consider the lack of visibility a problem (i.e. is this a network optimisation issue?).

> We introduce the field of Decentralized Data Delivery Markets (3DMs) to the world as a new field of research and innovation and one that has a rapidly increasing necessity of existence.  Decentralised storage networks, such as Filecoin, have reached unprecedented storage capacity commitments (of over 2.5EB of cold storage in Jan 2021) and continue to grow rapidly. There is an urgent need to complement decentralised storage with decentralised data delivery as these storage networks are seeking for a solution to deliver the data stored in their network to their end users, while meeting the expectations that end-users are used to with the centralised services of today.

Reviewer: ```suggestion
We introduce Decentralized Data Delivery Markets (3DMs) as a new field of research and innovation and one with rapidly increasing relevance.  Decentralised storage networks, such as Filecoin, have reached unprecedented storage capacity commitments (of over 2.5EB of cold storage in Jan 2021) and continue to grow rapidly. There is an urgent need to complement decentralised storage with decentralised data delivery as these storage networks seek a solution to deliver the data stored in their network to end-users while meeting the expectations users have of the centralised services of today.
```

We envision a 3DM as being:

*   A permissionless and open data delivery market, composed of one or more crypto-economic models
*   A fair exchange relationship between service providers and users
*   A cost efficient way for publishers to distribute their content
*   A unique set of business opportunities including:
    *   Delivering data to end users
    *   Carrying data between multiple disconnected locations or with high latency between them (Data Muling and/or Package Switching)

>     *   Create ephemeral distribution networks for large gatherings

Reviewer: ```suggestion
    *   Creating ephemeral distribution networks for large gatherings
```

>     *   Power the next generation of Web 3.0 applications

Reviewer: ```suggestion
    *   Powering the next generation of Web 3.0 applications
```

> This new field is ripe with unique business opportunities given the growing demand from users for access to large datasets, videos and so on, the panoply of super-powered mobile devices that end users have access to and the limitations imposed by physics in ultimately moving increasingly large data at high speeds through fiber.

Reviewer: ```suggestion
This new field is ripe with unique business opportunities given the growing demand from users for access to large datasets, videos, and other data-heavy content, the panoply of powerful mobile devices that users carry, and the limitations imposed by physics in ultimately moving increasingly large data at high speeds.
```

> In this Open Problem definition, we introduce  the multiple areas of research (or subOpenProblems), what we know so far for each one of them and link to some new ideas being explored.

Reviewer: Can we just call them **open sub-problems**? I think English would appreciate it 😄

### Definition of the actors

> For convenience and shared understanding, we define here the agents within a 3DM network. These are:
> 
> *   **Clients** - Fetching data from Providers
> *   **Providers** - Offer the service of data delivery to clients
> *   **Content Publishers** - The creators of content (or intermediaries) that want to have content distributed. There might be situations in which Content Publishers are willing to pay for the service consumed by Clients

Reviewer: ```suggestion
For convenience and shared understanding, we first define the agents within a 3DM network. These are:

*   **Clients** - Agents that fetch data from Providers.
*   **Providers** - Agents that offer data delivery services to Clients.
*   **Content Publishers** - Agents that create content and want to have it distributed (or intermediaries thereof). 
```

The part about publishers sometimes being willing to pay was out of place, as no one has talked about who pays at this point.

### Expected requirements

We present this list as a guide to protocol designers of what we expect a 3DM implementation to meet. This list is non-exhaustive and presents some flexibility between MUST and SHOULD. The requirements identified are:

*   **_MUST be Decentralized and not just federated_**
    *   Anyone should be able to join and leave at any time, without requiring permission
*   **_The exchanges of value_** **_MUST be verifiable_** 
    *   The payment for bandwidth/latency in token should match what has been agreed and fulfilled
    *   The payment should be fulfilled if the SLA is fulfilled

>     *   Parties should be able to verify that the service provided was correct (e.g. received the right file)

Reviewer: ```suggestion
    *   Parties should be able to verify that the service provided was correct (e.g. that they received the right file)
```

*   **_SHOULD be Trustless_**

>     *   Ideally, the network can perform the operations without having to leverage Trust in third parties and/or the participants of the exchange
>     *   Nevertheless, Trust can, and should, be considered as many designs might benefit from an element of Trust to increase the quality of service (e.g. reputation supports relaxing constraints)

Reviewer: ```suggestion
    *   Ideally, the network can perform the operations without having to leverage trust in third parties and/or the participants of the exchange
    *   Nevertheless, designs might benefit from an element of trust to increase the quality of service (e.g. reputation mechanisms may allow for relaxed constraints)
```

Written like that, it appears "Trust" is a specific concept in the context of the system, when it's just generally defined trust.

*   **_The network SHOULD do resource allocation in an efficient way_** (or as efficiently as possible)

>     *   Should not be an afterthought process

Reviewer: ```suggestion
    *   Should not be an afterthought
```

> *   **_Providers SHOULD coordinate and accept pre-fetching of content (warm up the catches)_**

Reviewer: ```suggestion
*   **_Providers SHOULD coordinate and accept pre-fetching of content (warm up the caches)_**
```

    *   This is in contrast to IPFS’s core principle of replicating only after explicit request by a peer
    *   This is done in order to make the network more efficient. However it is not mandatory with the setting being left up to the Provider.

Additional expectations (i.e. bonus points):

> *   Data SHOULD be readily available without having to wait for the [unsealing process](https://spec.filecoin.io/#section-glossary.seal) at storage nodes (e.g. Filecoin Storage Miners).

Reviewer: This is very Filecoin-specific. Could we instead write "without having to be retrieved from cold storage", with the unsealing process as an example?

> *   Data retrieval SHOULD be anonymous, privacy-preserving.

Reviewer: ```suggestion
*   Data retrieval SHOULD be anonymous and privacy-preserving.
```

> Other considerations (not requirements but good to think about):

Reviewer: ```suggestion
Other considerations:
```

*   It is expected that content will always be identified and retrieved by CIDs
*   The provider network will likely not provide random access to files, i.e., items will be named at the object level.

> *   In case there is an auction market, it should always be available (Liveness)
>     *   There needs to be a clever and fast way to match offers to bids

Reviewer: ```suggestion
*   In case there is an auction market, it should always be available and have a fast way to match offers to bids
```

# Areas of Work

> Throughout the exploration of the design space, we identified 3 (+1 bonus opportunity) areas of work that need to be explored in order to make 3DMs Fair, Decentralized and Efficient. The areas are listed below.

Reviewer: ```suggestion
Throughout the exploration of the design space, we identified 3 areas of work that need to be explored in order to make 3DMs Fair, Decentralized and Efficient, as well as a bonus opportunity. The areas are listed below.
```

#### Known challenges

*   Fairness
  *   Making it Trustless
    *   How to verify that the provider has the ability to delivery the file (e.g. stored locally or that can pull it from the network)
    *   How to verify that the file being transferred is the one requested, before a large amount of resources has been invested (i.e., before the entirety of the file has been transmitted)
    *   How to verify that the client has received the file
  *   How to verify and reward that SLA were guaranteed?
  *   How to avoid collusion when adding a third-party (e.g. Referee)?
  *   How to avoid grieving attacks (make one party pay for fees)?
  *   How to avoid a malicious actor causing un-rewarded work, hence wasting the others’ resources?
*   Experience
  *   How to make the transfers start instantaneously?
  *   How to support others (e.g. Content Publishers) paying for the usage?
  *    How to make it private (e.g. that others don’t know who is requesting what)?
*   Performance
  *   How to overcome the send-and-halt in order to max bandwidth throughput 
      *   How to support multipath (i.e. fetching from multiple sources)

> ### State-of-the-art

Author: - [ ] (@daviddias) Add descriptions to each of the areas + link to papers on their published venues

    *   Type: client fraud 
    *   Consequence: bandwidth-spent 
*   Malicious actor forces provider to pull the file from storage point (e.g. Filecoin, Cloud Storage, etc), paying the cost without the provider ever getting paid 
    *   Type: client fraud: 
    *   Consequence: currency is spent
*   Provider claims it has sent the file, but actually never did 
    *   Type: Provider fraud
    *   Consequence: Client gets penalized without receiving the service
*   Metering Inflation 
    *   Providers report to have used more resources to serve content to clients than what they have actually done.
*   Sybil / Throttle attacks
    *   Description: A set of nodes collude against a content-provider to try and make him go bankrupt (loss of all its “SLA stake”). They all try to download content from that content provider at the same time, so that Providers start consuming the stake from the content providers. If clients are not paying for the service we take the risk of enabling this attack. 

### New ideas being explored

ResNetLab organized a Research Intensive Workshop on 3DMs, out of which, the following ideas, structured as RFCs emerged:

* Link to RFCs:

>   * TOFILL

Author: Link to the following docs when they are merged:
- https://github.com/protocol/ResNetLab/pull/17
- https://github.com/protocol/ResNetLab/pull/16
- https://github.com/protocol/ResNetLab/pull/15

Properties

*   The system MUST always be able to discover content and satisfy content requests
    *   There should never be a “404 Content not Found” error.
*   The system MUST replicate content to different storage points in order to reduce delivery times and maximize performance.
*   Providers MUST follow the crypto-economic model and the system MUST make sure that Providers do not misbehave.
*   The network MUST be a content-addressable network and operate based on content identifiers.
*   The system MUST be permissionless and trustless
    *   Anyone should be free to join and set up a retrieval miner to contribute to the network.

> ### State-of-the-art

Author: - [ ] (@yiannisbot) Link to papers in their published venues

**Energy Efficiency and Battery Consumption**

Opportunistic communications that rely heavily (or exclusively) on user devices always come with the challenge of usability from the energy/battery perspective. Of course, incentives through FIL will alleviate some of it, but even then the incentive would have to be such that it overcomes the “running out of battery” shock. Techniques to overcome this can be as simple as disabling user participation when the battery level is below X%.

**Incentives**

The incentive to participate and the cryptoeconomic model behind the opportunistic part of the network is quite likely going to be incompatible with the one in the main Distribution Graph Forming area that applies to large/stable Providers. Nevertheless, incentives is a core part of this area and has traditionally been a stumbling block for the deployment of opportunistic networks.

### New ideas being explored

ResNetLab organized a Research Intensive Workshop on 3DMs, out of it, the following ideas, structured as RFCs emerged:

Link to RFCs

> *   TO FILL

Reviewer 2: Link to the following docs when they are merged:
* #18 
* #19 
* #20

*   Performance
  *   How to overcome the send-and-halt in order to max bandwidth throughput 
      *   How to support multipath (i.e. fetching from multiple sources)

### State-of-the-art

When it comes to the state of the art, we found that the existing solutions fall within one of the following categories:

#### Pay-per-packet

Pay-per-packet solutions offer granular control over what gets paid, enabling the option to verify if the SLA is being fulfilled. However, their setup cost and latency overhead puts them behind on being able to max out the available underlay resources.

These type of solutions have been found/highlighted:

- [Filecoin Retrieval Market](https://spec.filecoin.io/#section-systems.filecoin_markets.retrieval_market)
- [Theta Whitepaper](https://s3.us-east-2.amazonaws.com/assets.thetatoken.org/Theta-white-paper-latest.pdf?v=1612955013.791)

**Known shortcomings of these approaches:**
*   send-and-halt - Because the next packet (unit of service) will only be issued after the payment of the previous, these solutions often miss to max out the bandwidth available in the connection available, increasing the latency of delivery.

> #### Lock/Unlock access to the resource
> 
> INSERT DESCRIPTION
> 
> [(David) Solving the Buyer and Seller’s Dilemma: A Dual-Deposit Escrow Smart Contract for Provably Cheat-Proof Delivery and Payment for a Digital Good without a Trusted Mediator](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.nnf8jfjvpmgc)	[4](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.nnf8jfjvpmgc)
> [(David+Alfonso) FairSwap: How to fairly exchange digital goods](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.cbnzknlpc633)	[5](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.cbnzknlpc633)
> (David) Zero-Knowledge Contingent Payments Revisited: Attacks and Payments for Services	[5](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.it3yajc2rdz0)
> 
> **Known shortcomings of these approaches:**
> *   TOFILL

Author: ```suggestion
#### Lock/Unlock access to the resource

These type of solutions focus give the possibility for the client and the provider to verify the actual service before unlocking the payment. One of the main tools used is locking the access to the service by encrypting it with a key and then performing an exchange over the key that gives access to the service.

These solutions can be found in several forms:
- [Solving the Buyer and Seller’s Dilemma: A Dual-Deposit Escrow Smart Contract for Provably Cheat-Proof Delivery and Payment for a Digital Good without a Trusted Mediator](INSERT URL)
- [FairSwap: How to fairly exchange digital goods](INSERT URL)
- [Zero-Knowledge Contingent Payments Revisited: Attacks and Payments for Services](INSERT URL)	

**Known shortcomings of these approaches:**
*   Vulnerable to Grieving attacks.
*   Generally, these solutions focus on the fulfilment of the delivery, more so than measuring the speed and rate of the transfer.

```

*   Vulnerable to Grieving attacks.
*   Generally, these solutions focus on the fulfilment of the delivery, more so than measuring the speed and rate of the transfer.

#### Optimistic Fair Exchange

These solutions build on top of the Fair Exchange ones and adopt an Optimistic approach where some sort of Reputation or Stake is used to give both parties trust that the exchange will occur correctly. If one misbehaves, they can still rely on a dispute system.

- [OptiSwap: Fast Optimistic Fair Exchange](INSERT URL)
- [Asynchronous Protocols for Optimistic Fair Exchange](INSERT URL)
- [Optimistic Fair Exchange with Multiple Arbiters](INSERT URL)

**Known shortcomings of these approaches:**
*   Setup + Dispute resolution time is non insignificant and non-trivial

#### Reputation based

Reputation based solutions give the possibility to speed up the transfer and improve the quality of the service, by leveraging the trust built over time between clients & providers.

> - [(David) Proof of Delivery in a Trustless Network](INSERT URL)
> - [(David) Proof-of-Prestige: A Useful Work Reward System for Unverifiable Tasks](INSERT URL)
> - [(David) Mechanisms for Outsourcing Computation via a Decentralized Market](INSERT URL)
> - Other Gradual Fair Exchange Class of constructions

Author: ```suggestion
- [Proof of Delivery in a Trustless Network](INSERT URL)
- [Proof-of-Prestige: A Useful Work Reward System for Unverifiable Tasks](INSERT URL)
- [Mechanisms for Outsourcing Computation via a Decentralized Market](INSERT URL)
- Other Gradual Fair Exchange Class of constructions
```

**Known shortcomings of these approaches:**
*   Not a direct shortcoming, but a challenge is that each participant my have different levels of sensitivity with regards to how much they are willing to rely on trust vs. strict verifiability. 

#### Privacy focused

There are multiple solutions with the goal of delivering readers privacy and writers privacy. These incur an additional cost in setup and/or latency in favor of the additional privacy preserving benefits. Disclaimer: At the time of writing, we haven't dived as deep into these set of solutions.

- [Blockchain based Privacy-Preserving Software Updates with Proof-of-Delivery for Internet of Things](INSERT URL)
- [SilentDelivery: Practical Timed-delivery of Private Information using Smart Contracts](INSERT URL)
- [Bulletproofs+: Shorter Proofs for Privacy-Enhanced Distributed Ledger](INSERT URL)

#### Traditional approaches (close to Web 2.0 world)

> INSERT DESCRIPTION
> 
> [(David) The multimedia blockchain: a distributed and tamper-proof media transaction framework](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.tsv8bvhhclyh)	[10](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.tsv8bvhhclyh)
> [(Alfonso) Reliable Client Accounting for P2P-Infrastructure Hybrids](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.mnagl72zaw3u)	[11](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.mnagl72zaw3u)
> 
> **Known shortcomings of these approaches:**
> *   TOFILL

Author: ```suggestion

These approaches suggest novel ways to add integrity checks to data transferred or to promote new ways for content distribution, but they don’t depart from the Web 2.0 world of assumptions such as trusted third parties and central points of control.

- [The multimedia blockchain: a distributed and tamper-proof media transaction framework](INSERT URL)
- [Reliable Client Accounting for P2P-Infrastructure Hybrids](INSERT URL)

**Known shortcomings of these approaches:**
*   Not fully aware of the challenges of building a decentralized network
```

#### Properties desired

*   The exchanges of value MUST be verifiable and correct
    *   Fairness: 
        *   The payment MUST be fulfilled if the SLA is fulfilled
        *   The payment for bandwidth/latency in token SHOULD match what has been agreed and fulfilled
    *   Verifiability:
        *   Both parties MUST be capable of verifying that the exchange was performed correctly
        *   Bonus: Anyone SHOULD be able to verify that the exchange was performed correctly
*   SHOULD be Trustless
    *   Ideally, the network can perform the operations without having to leverage Trust
    *   Nevertheless in many designs it might be needed to leverage reputation as a trust vehicle to guarantee high quality of service and/or reliability.

#### How is Data Delivery Metering different from Fair Exchange

The field of Fair Exchange overlaps in part with Service Metering. Some notable differences are:

> *   Fair Exchange can be used for non-digital goods also

Reviewer: I'm unsure about this distinction. Metering can also be used for petrol 🤷 Not in this particular context, perhaps, but the same is true of fair exchange.

*   Fair Exchange is traditionally between 2 parties
*   Metering should be available to properly measure the service quality and service provided by one or more parties (e.g. streaming from multiple endpoints)

In review, the Metering & Fair Exchange fields end up contributing to each other and it is likely that the intersection of both is needed for a sound solution. 

#### Known challenges

*   Fairness
  *   Making it Trustless

>     *   How to verify that the provider has the ability to delivery the file (e.g. stored locally or that can pull it from the network)

Reviewer: ```suggestion
    *   How to verify that the provider has the ability to delivery the file (e.g. has a local copy or is able to pull it from the network)
```

>     *   How to verify that the file being transferred is the one requested, before a large amount of resources has been invested (i.e., before the entirety of the file has been transmitted)

Reviewer: ```suggestion
    *   How to verify that the file being transferred is the one requested, before a large amount of resources has been invested (i.e. before the entirety of the file has been transmitted)
```

    *   How to verify that the client has received the file

>   *   How to verify and reward that SLA were guaranteed?

Reviewer: ```suggestion
  *   How to verify and reward that SLA were met?
```

  *   How to avoid collusion when adding a third-party (e.g. Referee)?

>   *   How to avoid grieving attacks (make one party pay for fees)?

Reviewer: ```suggestion
  *   How to avoid griefing attacks (make one party pay for fees)?
```

>   *   How to avoid a malicious actor causing un-rewarded work, hence wasting the others’ resources?

Reviewer: ```suggestion
  *   How to avoid a malicious actor causing un-rewarded work, hence wasting others’ resources?
```

*   Experience
  *   How to make the transfers start instantaneously?
  *   How to support others (e.g. Content Publishers) paying for the usage?

>   *    How to make it private (e.g. that others don’t know who is requesting what)?

Reviewer: ```suggestion
  *    How to make it private (i.e. so that others don’t know which users are requesting what content)?
```

*   Performance

>   *   How to overcome the send-and-halt in order to max bandwidth throughput 

Reviewer: ```suggestion
  *   How to overcome send-and-halt in order to maximize bandwidth throughput 
```

I'm not sure what this means. By "send-and-halt", are you referring to block-by-block request and retrieval loops?

Reviewer: Question answered below. Change suggested should be fine.

      *   How to support multipath (i.e. fetching from multiple sources)

### State-of-the-art

When it comes to the state of the art, we found that the existing solutions fall within one of the following categories:

#### Pay-per-packet

> Pay-per-packet solutions offer granular control over what gets paid, enabling the option to verify if the SLA is being fulfilled. However, their setup cost and latency overhead puts them behind on being able to max out the available underlay resources.

Reviewer: ```suggestion
Pay-per-packet solutions offer granular control over what gets paid, enabling the option to verify if the SLA is being fulfilled. However, their setup cost and latency overhead put them at a disadvantage in terms of maximizing utilization of the underlying resources.
```
I _think_ you meant underlying instead of underlay but revert if not.

These type of solutions have been found/highlighted:

- [Filecoin Retrieval Market](https://spec.filecoin.io/#section-systems.filecoin_markets.retrieval_market)
- [Theta Whitepaper](https://s3.us-east-2.amazonaws.com/assets.thetatoken.org/Theta-white-paper-latest.pdf?v=1612955013.791)

**Known shortcomings of these approaches:**

> *   send-and-halt - Because the next packet (unit of service) will only be issued after the payment of the previous, these solutions often miss to max out the bandwidth available in the connection available, increasing the latency of delivery.

Reviewer: ```suggestion
*   send-and-halt - Because the next packet (unit of service) will only be issued after payment for the previous is received, these solutions often fail to max out the bandwidth available in the connection, increasing the latency of delivery.
```

#### Lock/Unlock access to the resource

> These type of solutions focus give the possibility for the client and the provider to verify the actual service before unlocking the payment. One of the main tools used is locking the access to the service by encrypting it with a key and then performing an exchange over the key that gives access to the service.

Reviewer: ```suggestion
Solutions of this type enable the client and the provider to verify the actual service before unlocking the payment. One of the main tools used is locking the access to the service by encrypting it with a key and then performing an exchange of the key that gives access to the service.
```

These solutions can be found in several forms:
- [Solving the Buyer and Seller’s Dilemma: A Dual-Deposit Escrow Smart Contract for Provably Cheat-Proof Delivery and Payment for a Digital Good without a Trusted Mediator](INSERT URL)
- [FairSwap: How to fairly exchange digital goods](INSERT URL)
- [Zero-Knowledge Contingent Payments Revisited: Attacks and Payments for Services](INSERT URL)	

**Known shortcomings of these approaches:**

> *   Vulnerable to Grieving attacks.

Reviewer: ```suggestion
*   Vulnerable to griefing attacks.
```

*   Generally, these solutions focus on the fulfilment of the delivery, more so than measuring the speed and rate of the transfer.

#### Optimistic Fair Exchange

These solutions build on top of the Fair Exchange ones and adopt an Optimistic approach where some sort of Reputation or Stake is used to give both parties trust that the exchange will occur correctly. If one misbehaves, they can still rely on a dispute system.

- [OptiSwap: Fast Optimistic Fair Exchange](INSERT URL)
- [Asynchronous Protocols for Optimistic Fair Exchange](INSERT URL)
- [Optimistic Fair Exchange with Multiple Arbiters](INSERT URL)

**Known shortcomings of these approaches:**

> *   Setup + Dispute resolution time is non insignificant and non-trivial

Reviewer: ```suggestion
*   Setup + Dispute resolution is non-trivial and takes significant time
```

#### Reputation based

> Reputation based solutions give the possibility to speed up the transfer and improve the quality of the service, by leveraging the trust built over time between clients & providers.

Reviewer: ```suggestion
Reputation-based solutions give the possibility to speed up the transfer and improve the quality of the service by leveraging the trust built over time between clients & providers.
```

- [Proof of Delivery in a Trustless Network](INSERT URL)
- [Proof-of-Prestige: A Useful Work Reward System for Unverifiable Tasks](INSERT URL)
- [Mechanisms for Outsourcing Computation via a Decentralized Market](INSERT URL)
- Other Gradual Fair Exchange Class of constructions

**Known shortcomings of these approaches:**
*   Not a direct shortcoming, but a challenge is that each participant my have different levels of sensitivity with regards to how much they are willing to rely on trust vs. strict verifiability. 

#### Privacy focused

> There are multiple solutions with the goal of delivering readers privacy and writers privacy. These incur an additional cost in setup and/or latency in favor of the additional privacy preserving benefits. Disclaimer: At the time of writing, we haven't dived as deep into these set of solutions.

Reviewer: ```suggestion
There are multiple solutions with the goal of delivering reader privacy and writer privacy. These incur an additional cost in setup and/or latency. Disclaimer: At the time of writing, we haven't dived deep into this set of solutions.
```

- [Blockchain based Privacy-Preserving Software Updates with Proof-of-Delivery for Internet of Things](INSERT URL)
- [SilentDelivery: Practical Timed-delivery of Private Information using Smart Contracts](INSERT URL)
- [Bulletproofs+: Shorter Proofs for Privacy-Enhanced Distributed Ledger](INSERT URL)

#### Traditional approaches (close to Web 2.0 world)

These approaches suggest novel ways to add integrity checks to data transferred or to promote new ways for content distribution, but they don’t depart from the Web 2.0 world of assumptions such as trusted third parties and central points of control.

- [The multimedia blockchain: a distributed and tamper-proof media transaction framework](INSERT URL)
- [Reliable Client Accounting for P2P-Infrastructure Hybrids](INSERT URL)

**Known shortcomings of these approaches:**
*   Not fully aware of the challenges of building a decentralized network

### Known attacks to be mitigated

Here we list the attacks to consider when designing a solution. These are:

> *   Malicious actor forces provider to spend bandwidth without issuing payment (also known as Grieving Attack)

Reviewer: ```suggestion
*   Malicious actor forces provider to spend bandwidth without issuing payment (also known as Griefing Attack)
```
I don't appreciate the corruption either but it really is grief/griefing attack 🤷

    *   Type: client fraud 

>     *   Consequence: bandwidth-spent 

Reviewer: ```suggestion
    *   Consequence: bandwidth is spent 
```

> *   Malicious actor forces provider to pull the file from storage point (e.g. Filecoin, Cloud Storage, etc), paying the cost without the provider ever getting paid 

Reviewer: ```suggestion
*   Malicious actor forces provider to pay to retrieve the file from storage point (e.g. Filecoin, Cloud Storage, etc) without the provider itself ever getting paid 
```

>     *   Type: client fraud: 

Reviewer: ```suggestion
    *   Type: client fraud
```

    *   Consequence: currency is spent
*   Provider claims it has sent the file, but actually never did 
    *   Type: Provider fraud

>     *   Consequence: Client gets penalized without receiving the service

Reviewer: Penalized why/how?

*   Metering Inflation 
    *   Providers report to have used more resources to serve content to clients than what they have actually done.
*   Sybil / Throttle attacks

>     *   Description: A set of nodes collude against a content-provider to try and make him go bankrupt (loss of all its “SLA stake”). They all try to download content from that content provider at the same time, so that Providers start consuming the stake from the content providers. If clients are not paying for the service we take the risk of enabling this attack. 

Reviewer: ```suggestion
    *   Description: A set of nodes collude against a content publisher to try and make it go bankrupt (i.e. trigger a loss of all its “SLA stake”). The colluding nodes simultaneously try to download content offered by a content publisher in a way that causes providers to start consuming the publisher's stake. This attack is possible if clients are not required to pay for service. 
```
I think this is what you meant (i.e. "content provider" should be "content publisher").

    *   It should also avoid “malicious economic attacks” such as a content provider trying to bankrupt its competition by draining the stake of his retrieval deal.
*   The model SHOULD foster collaboration between entities to benefit users perceived QoS / QoE. 
    *   The model should reward entities that collaborate to serve content efficiently and with high QoS to users.
    *   This will also result in an efficient allocation of resources in the network.
*   If any part of the economics of the system ends up being orchestrated by an auction market, this market MUST be always available.
    *   And should match offers to bids in a fast and efficient way.

### State-of-the-art

As described above, the outcome of this topic should be to design an economic model that aligns the goals of all profit-maximizing entities in the system while preventing misbehaviors. Given this target, we split our investigation in the following areas:

> ##### Economics of Hybrid CDNs and P2P networks

Reviewer: Remember to link papers here too

The economics of current Hybrid CDN deployments give a good understanding of the system and cost models involved in the delivery of content at scale. This, added to the different analysis and design proposals of economic models for content delivery in P2P networks represent a good foundation to understand the complexity of the problem and potential ways to tackle it.

**_Related Papers:_**

*   An economic mechanism for request routing and resource allocation in hybrid CDN–P2P networks
*   An economic replica placement mechanism for streaming content distribution in Hybrid CDN-P2P networks
*   How Neutral is a CDN? An Economic Approach
*   Value Networks and Two-Sided Markets of Internet Content Delivery
*   Bilateral and Multilateral Exchanges for Peer-Assisted Content Distribution

**_Learnings: _**

*   Protocols and systems involved in the cost model of a Hybrid-CDN.
*   Models of P2P content delivery networks.

##### Game theoretical models and reward/reputation systems in P2P networks

> How to design the right incentive and reputation models to incentivize certain behaviors in P2P networks has been a widely studied problem for some time now. From reputation systems where nodes track the level of good behavior of their peers and cooperate with them according to their grade; to incentive systems where good behavior is explicitly rewarded. To theoretically evaluate the strength of these schemes, they are generally mod game theoretical models are generally an objective way to evaluate the strength of these schemes, and to understand if the proposals can achieve Nash Equilibrium.

Reviewer: ```suggestion
How to design the right incentive and reputation models to incentivize certain behaviors in P2P networks has been a widely studied problem for some time now. From reputation systems where nodes track the level of good behavior of their peers and cooperate with them according to their grade; to incentive systems where good behavior is explicitly rewarded. Game-theoretical models are a way to evaluate the strength of these schemes and to understand if the proposals can achieve Nash Equilibrium.
```

We call these environments “opportunistic deployments” to reflect their unpredictability in terms of availability, uptime, resource quality and quantity. In contrast to RMs, as defined earlier that are expected to have stable, public and high-bandwidth connectivity, 3DM opportunistic deployments can utilise everyday user devices to extend the storage footprint of RMs and create a wealth of new network connectivity and business opportunities.

Although this area seems to sit on the periphery of 3DMs, it is actually a highly impactful area, as it realises the vision of Decentralised Storage Networks in general and Filecoin, in particular, of regular end-users sharing their own resources to contribute to the network. That said, we place high value in capturing this opportunity and offering to end-users the opportunity of being rewarded for their contribution to the network.

### State-of-the-art

The literature in the field of Opportunistic and Delay-Tolerant Networks is vast. We have narrowed down the scope of the literature that we reviewed and identified the following most-promising 3DM applicability areas.

#### Transient Providers

> We consider end-users that store content in their ephemerally connected devices (laptops, or mobile phones) and provide them to the network through (one or more) Providers, as defined in the “Distribution Graph Forming” area. These service providers are called “Transient Providers (TRMs) to depict their ephemeral nature in resource availability and time. TRMs can have some economic relationship with one or more RMs, e.g., RMs can “recruit” TRMs to expand their storage capacity and footprint. In this case, RMs have to maintain their own monitoring mechanisms and do resource allocation to ensure that they are getting desirable levels of service from their group of TRMs.

Reviewer: ```suggestion
We consider end-users that store content in their ephemerally connected devices (laptops, or mobile phones) and provide them to the network through (one or more) Providers, as defined in the “Distribution Graph Forming” area. These service providers are called Transient Providers (TRMs) to depict their ephemeral nature in resource availability and time. TRMs can have some economic relationship with one or more RMs, e.g., RMs can “recruit” TRMs to expand their storage capacity and footprint. In this case, RMs have to maintain their own monitoring mechanisms and do resource allocation to ensure that they are getting desirable levels of service from their group of TRMs.
```
Do we want to keep the TRM initialism (from transient **retrieval miners**) even though we refer to them as transient **providers**?

Opportunistic D2D

This area of research and deployment is closer to traditional Delay-Tolerant Networks (DTNs). We envision applications where mobile devices contribute to the distribution of content in the mobile domain. These can be akin to previously proposed [User-Operated Mobile Content Distribution Networks](https://drive.google.com/file/d/1vvtQ7MJb4YFRXxeo2GewGh8EVYEqV-iJ/view?usp=sharing), or applications to realise concepts such as “[Floating Content](https://drive.google.com/file/d/1yOniHbAjLYv3q09pff7gnulOwT_NYCYG/view?usp=sharing)”. 

### Known shortcomings 

**User Mobility**

> When a user is “registered” under one address/network location to serve content (at least as far as its Retrieval Miner is concerned) and then moves to another network location, the content is not possible to find anymore.

Reviewer: ```suggestion
When a user is “registered” under one address/network location to serve content (at least as far as its Retrieval Miner is concerned) and then moves to another network location, the content is not discoverable/retrievable anymore.
```
There are still a handful of occurrences of "retrieval miner" in the text. I don't think the term is explained anywhere. Should all of these become "provider"?

#### Lock/Unlock access to the resource

These type of solutions focus give the possibility for the client and the provider to verify the actual service before unlocking the payment. One of the main tools used is locking the access to the service by encrypting it with a key and then performing an exchange over the key that gives access to the service.

These solutions can be found in several forms:
- [Solving the Buyer and Seller’s Dilemma: A Dual-Deposit Escrow Smart Contract for Provably Cheat-Proof Delivery and Payment for a Digital Good without a Trusted Mediator](INSERT URL)
- [FairSwap: How to fairly exchange digital goods](INSERT URL)
- [Zero-Knowledge Contingent Payments Revisited: Attacks and Payments for Services](INSERT URL)	

**Known shortcomings of these approaches:**
*   Vulnerable to Grieving attacks.
*   Generally, these solutions focus on the fulfilment of the delivery, more so than measuring the speed and rate of the transfer.

> #### Optimistic Fair Exchange
> 
> INSERT DESCRIPTION
> 
> [(David) OptiSwap: Fast Optimistic Fair Exchange](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.f2mby9dpar4f)	[6](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.f2mby9dpar4f)
> [(Irene) Asynchronous Protocols for Optimistic Fair Exchange](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.6ac4zrrdyhqs)	[6](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.6ac4zrrdyhqs)
> [(Irene) Optimistic Fair Exchange with Multiple Arbiters](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.sjnrx6peoi0w)	[7](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.sjnrx6peoi0w)
> 
> **Known shortcomings of these approaches:**
> *   TOFILL

Author: ```suggestion
#### Optimistic Fair Exchange

These solutions build on top of the Fair Exchange ones and adopt an Optimistic approach where some sort of Reputation or Stake is used to give both parties trust that the exchange will occur correctly. If one misbehaves, they can still rely on a dispute system.

- [OptiSwap: Fast Optimistic Fair Exchange](INSERT URL)
- [Asynchronous Protocols for Optimistic Fair Exchange](INSERT URL)
- [Optimistic Fair Exchange with Multiple Arbiters](INSERT URL)

**Known shortcomings of these approaches:**
*   Setup + Dispute resolution time is non insignificant and non-trivial
```

> #### Reputation based
> 
> INSERT DESCRIPTION
> 
> [(David) Proof of Delivery in a Trustless Network](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.g219mq8bs8ad)	[7](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.g219mq8bs8ad)
> [(David) Proof-of-Prestige: A Useful Work Reward System for Unverifiable Tasks](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.9sqsu17nv29v)	[8](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.9sqsu17nv29v)
> [(David) Mechanisms for Outsourcing Computation via a Decentralized Market](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.8lcwzhd2jj73)	[9](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.8lcwzhd2jj73)
> [(Irene) Gradual Fair Exchange Class of Constructions](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.nq0pu1otlwlb)	[10](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.nq0pu1otlwlb)
> 
> **Known shortcomings of these approaches:**
> *   TOFILL

Author: ```suggestion
#### Reputation based

Reputation based solutions give the possibility to speed up the transfer and improve the quality of the service, by leveraging the trust built over time between clients & providers.

- [(David) Proof of Delivery in a Trustless Network](INSERT URL)
- [(David) Proof-of-Prestige: A Useful Work Reward System for Unverifiable Tasks](INSERT URL)
- [(David) Mechanisms for Outsourcing Computation via a Decentralized Market](INSERT URL)
- Other Gradual Fair Exchange Class of constructions

**Known shortcomings of these approaches:**
*   Not a direct shortcoming, but a challenge is that each participant my have different levels of sensitivity with regards to how much they are willing to rely on trust vs. strict verifiability. 
```

> #### Privacy focused
> 
> INSERT DESCRIPTION
> 
> [(David) Blockchain based Privacy-Preserving Software Updates with Proof-of-Delivery for Internet of Things](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.3n5ydbvkkmoc)	[10](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.3n5ydbvkkmoc)
> [(David) SilentDelivery: Practical Timed-delivery of Private Information using Smart Contracts](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.2q9trcjswmyw)	[10](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.2q9trcjswmyw)
> [(David) Bulletproofs+: Shorter Proofs for Privacy-Enhanced Distributed Ledger](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.bweiezm16dvh)	[10](https://docs.google.com/document/d/1hXTP0VRmQsSFcnbxk3Z4sVmdW0CPajogJ6SXTe7zJ34/edit#heading=h.bweiezm16dvh)
> 
> **Known shortcomings of these approaches:**
> *   TBD

Author: ```suggestion
#### Privacy focused

There are multiple solutions with the goal of delivering readers privacy and writers privacy. These incur an additional cost in setup and/or latency in favor of the additional privacy preserving benefits. Disclaimer: At the time of writing, we haven't dived as deep into these set of solutions.

- [Blockchain based Privacy-Preserving Software Updates with Proof-of-Delivery for Internet of Things](INSERT URL)
- [SilentDelivery: Practical Timed-delivery of Private Information using Smart Contracts](INSERT URL)
- [Bulletproofs+: Shorter Proofs for Privacy-Enhanced Distributed Ledger](INSERT URL)
```

    *   Type: client fraud
    *   Consequence: bandwidth is spent
*   Malicious actor forces provider to pay to retrieve the file from storage point (e.g. Filecoin, Cloud Storage, etc) without the provider itself ever getting paid
    *   Type: client fraud
    *   Consequence: currency is spent
*   Provider claims it has sent the file, but actually never did
    *   Type: Provider fraud
    *   Consequence: Client gets penalized without receiving the service
*   Metering Inflation
    *   Providers report to have used more resources to serve content to clients than what they have actually done.
*   Sybil / Throttle attacks
    *   Description: A set of nodes collude against a content-provider to try and make him go bankrupt (loss of all its “SLA stake”). They all try to download content from that content provider at the same time, so that Providers start consuming the stake from the content providers. If clients are not paying for the service we take the risk of enabling this attack.

### New ideas being explored

ResNetLab organized a Research Intensive Workshop on 3DMs, out of which, the following ideas, structured as RFCs emerged:

> * Link to RFCs:
>   * TOFILL

Author: ```suggestion
* [RFC: ZKCP with Fair Exchanges of 1 bit](https://github.com/protocol/ResNetLab/blob/master/3DM_RFC/RIW2021_RFC_ZKCP%20with%20Fair%20Exchanges%20of%201%20bit.md)
* [RFC: ZKCP Optimizations](https://github.com/protocol/ResNetLab/blob/master/3DM_RFC/RIW2021_RFC_ZKCP%20Optimizations.md)
* [RFC_Optimistic ZKC(S)P](https://github.com/protocol/ResNetLab/pull/15)
```

    *   The payment for bandwidth/latency in token should match what has been agreed and fulfilled
    *   The payment should be fulfilled if the SLA is fulfilled
    *   Parties should be able to verify that the service provided was correct (e.g. that they received the right file)
*   **_SHOULD be Trustless_**
    *   Ideally, the network can perform the operations without having to leverage trust in third parties and/or the participants of the exchange
    *   Nevertheless, designs might benefit from an element of trust to increase the quality of service (e.g. reputation mechanisms may allow for relaxed constraints)
*   **_The network SHOULD do resource allocation in an efficient way_** (or as efficiently as possible)
    *   Should not be an afterthought
*   **_Providers SHOULD coordinate and accept pre-fetching of content (warm up the caches)_**
    *   This is in contrast to IPFS’s core principle of replicating only after explicit request by a peer
    *   This is done in order to make the network more efficient. However it is not mandatory with the setting being left up to the Provider.

Additional expectations (i.e. bonus points):

> *   Data SHOULD be readily available without having to wait for the [unsealing process](https://spec.filecoin.io/#section-glossary.seal) at storage nodes (e.g. Filecoin Storage Miners).

Author: ```suggestion
```

In a decentralized environment that strives to be permissionless, we can’t expect to rely on such legal agreements and third parties and, therefore, need a way to prove fair exchange. That is, we need to be able to prove that the service was delivered correctly and the metering of the provided service was done correctly. This will, in turn, verify the exchange between the two parties and issue the correct payment (i.e. fair exchange).

This trustless property is really important as we know that markets with no mediators, escrows, and other assurance services (e.g. underground markets) are prone to scams, as the users have to trust that the provider will execute on the agreement. In a trustless environment, once the transfer is completed, there is no way to dispute it.

#### Properties desired

> *   The exchanges of value MUST be verifiable and correct

Author: ```suggestion
*   The exchanges of value MUST be verifiable and fair
```
